File: app/src/main/java/ardjomand/leonardo/nutrimeal/companyorders/CompanyOrdersFragment.java
package ardjomand.leonardo.nutrimeal.companyorders;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.ArrayList;

import ardjomand.leonardo.nutrimeal.R;
import ardjomand.leonardo.nutrimeal.data.pojos.CompanyOrder;
import ardjomand.leonardo.nutrimeal.widget.OrdersWidget;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class CompanyOrdersFragment extends Fragment implements
        CompanyOrdersContract.View<CompanyOrder>,
        CompanyOrdersAdapter.OnMealAdapterInteractionListener {

    private static final String ARG_COLUMN_COUNT = "column-count";

    //region Views
    @BindView(R.id.list)
    RecyclerView recyclerView;
    //endregion

    //region Member variables
    private int mColumnCount = 1;
    private Unbinder unbinder;
    private CompanyOrdersContract.Presenter presenter;
    private CompanyOrdersAdapter adapter;
    //endregion

    //region Constructors
    public CompanyOrdersFragment() {
    }

    @SuppressWarnings("unused")
    public static CompanyOrdersFragment newInstance(int columnCount) {
        CompanyOrdersFragment fragment = new CompanyOrdersFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_COLUMN_COUNT, columnCount);
        fragment.setArguments(args);
        return fragment;
    }
    //endregion

    //region Lifecycle
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            mColumnCount = getArguments().getInt(ARG_COLUMN_COUNT);
        }

        presenter = new CompanyOrdersPresenter(this);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_orders, container, false);
        unbinder = ButterKnife.bind(this, view);

        setTitle();

        // Set layout manager
        Context context = view.getContext();
        if (mColumnCount <= 1) {
            recyclerView.setLayoutManager(new LinearLayoutManager(context));
        } else {
            recyclerView.setLayoutManager(new GridLayoutManager(context, mColumnCount));
        }

        // Set adapter
        adapter = new CompanyOrdersAdapter(new ArrayList<CompanyOrder>(), this, getContext());
        recyclerView.setAdapter(adapter);

        // Set decoration
        RecyclerView.ItemDecoration itemDecoration = new
                DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL);
        recyclerView.addItemDecoration(itemDecoration);

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        adapter.clearData();
        presenter.subscribe();
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.unsubscribe();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.setView(null);
    }

    //endregion

    //region Presenter callbacks
    @Override
    public void addItem(CompanyOrder companyOrder) {
        adapter.addData(companyOrder);
        OrdersWidget.updateAllWidgets(getContext());
    }

    @Override
    public void updateItem(CompanyOrder companyOrder) {
        adapter.updateData(companyOrder);
        OrdersWidget.updateAllWidgets(getContext());
    }

    @Override
    public void showEmptyItems() {
        Toast.makeText(getActivity(), R.string.error_items_could_not_be_downloaded, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showError() {
        Toast.makeText(getActivity(), R.string.error_no_items_available, Toast.LENGTH_SHORT).show();
    }
    //endregion

    // TODO move the responsibility of setting title to the parent activity
    // this will make it easier for the fragments to be added instead of replaced
    private void setTitle() {
        if (getActivity() instanceof AppCompatActivity) {
            ActionBar supportActionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
            if (supportActionBar != null) {
                supportActionBar.setTitle(R.string.orders_title);
                supportActionBar.setDisplayHomeAsUpEnabled(true);
            }
        }
    }

    @Override
    public void onOrderClicked(CompanyOrder companyOrder) {
        // TODO Add possibility to change order status
    }
}

File: app/src/main/java/ardjomand/leonardo/nutrimeal/companyorders/CompanyOrdersFragment.kt
package ardjomand.leonardo.nutrimeal.companyorders

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import ardjomand.leonardo.nutrimeal.R
import ardjomand.leonardo.nutrimeal.data.pojos.CompanyOrder
import ardjomand.leonardo.nutrimeal.widget.OrdersWidget
import kotlinx.android.synthetic.main.fragment_orders.*
import java.util.*

class CompanyOrdersFragment : Fragment(),
        CompanyOrdersContract.View<CompanyOrder>,
        CompanyOrdersAdapter.OnMealAdapterInteractionListener {

    //region Member variables
    private var columnCount = 1
    private lateinit var presenter: CompanyOrdersContract.Presenter
    private lateinit var adapter: CompanyOrdersAdapter
    //endregion

    //region Lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?.let {
            columnCount = it.getInt(ARG_COLUMN_COUNT)
        }

        presenter = CompanyOrdersPresenter(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_orders, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setTitle()

        // Set layout manager
        val context = view.context
        list.layoutManager = when {
            columnCount <= 1 -> LinearLayoutManager(context)
            else -> GridLayoutManager(context, columnCount)
        }

        // Set adapter
        adapter = CompanyOrdersAdapter(ArrayList(), this, getContext())
        list.adapter = adapter

        // Set decoration
        val itemDecoration = DividerItemDecoration(list.context, DividerItemDecoration.VERTICAL)
        list.addItemDecoration(itemDecoration)

    }

    override fun onStart() {
        super.onStart()
        adapter.clearData()
        presenter.subscribe()
    }

    override fun onStop() {
        super.onStop()
        presenter.unsubscribe()
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.setView(null)
    }

    //endregion

    //region Presenter callbacks
    override fun addItem(companyOrder: CompanyOrder) {
        adapter.addData(companyOrder)
        OrdersWidget.updateAllWidgets(context)
    }

    override fun updateItem(companyOrder: CompanyOrder) {
        adapter.updateData(companyOrder)
        OrdersWidget.updateAllWidgets(context)
    }

    override fun showEmptyItems() {
        Toast.makeText(activity, R.string.error_items_could_not_be_downloaded, Toast.LENGTH_SHORT).show()
    }

    override fun showError() {
        Toast.makeText(activity, R.string.error_no_items_available, Toast.LENGTH_SHORT).show()
    }
    //endregion

    // TODO move the responsibility of setting title to the parent activity
    // this will make it easier for the fragments to be added instead of replaced
    private fun setTitle() {
        if (activity is AppCompatActivity) {
            val supportActionBar = (activity as AppCompatActivity).supportActionBar
            supportActionBar?.apply {
                setTitle(R.string.orders_title)
                setDisplayHomeAsUpEnabled(true)
            }
        }
    }

    override fun onOrderClicked(companyOrder: CompanyOrder) {
        // TODO Add possibility to change order status
    }

    companion object {

        private const val ARG_COLUMN_COUNT = "column-count"

        fun newInstance(columnCount: Int) =
                CompanyOrdersFragment().apply {
                    arguments = Bundle().apply {
                        putInt(ARG_COLUMN_COUNT, columnCount)
                    }
                }
    }
}
